### PR TITLE
Add a throttle for pending challenges

### PR DESCRIPTION
**Pull request by mia-pi-git**
This is to resolve an issue where a user would be spammed by challenges from unregged bot accounts until their browser crashed. 
This PR:
- prevents unnamed users from challenging 
- Prevents a user from being challenged by a non-autoconfirmed user if they already have 3 challenges pending.
In short, should stop users from spambot-challenging.

### CODE REVIEW

File: server/users.ts [hunk 1]
```diff
@@ -382,6 +382,7 @@ export class User extends Chat.MessageContext {
 	statusType: StatusType;
 	userMessage: string;
 	lastWarnedAt: number;
+	currentChallengeCount: number;
```
Comment (Zarel) [inline server/users.ts:385]: I don't think this needs more state; the challenge count should be able to be calculated from existing state (`Ladders.challenges`?)

File: server/ladders.ts [hunk 2]
```diff
@@ -296,7 +304,9 @@ class Ladder extends LadderStore {
 			const fromUser = Users.get(challenge.from);
 			if (fromUser) Ladder.updateChallenges(fromUser);
 			const toUser = Users.get(challenge.to);
-			if (toUser) Ladder.updateChallenges(toUser);
+			if (toUser) {
+				Ladder.updateChallenges(toUser);
+			}
```
Comment (Zarel) [inline server/ladders.ts:309]:
```suggestion
			if (toUser) Ladder.updateChallenges(toUser);
```

File: server/ladders.ts [hunk 3]
```diff
@@ -314,7 +324,9 @@ class Ladder extends LadderStore {
...
```
Comment (Zarel) [inline server/ladders.ts:329]:
```suggestion
			if (toUser) Ladder.updateChallenges(toUser);
```

File: server/chat-commands/core.ts [hunk 4]
```diff
@@ -1237,6 +1237,9 @@ export const commands: ChatCommands = {
 		if (Punishments.isBattleBanned(user)) {
 			return this.popupReply(`You are banned from battling and cannot challenge users.`);
 		}
+		if (!user.named || user.namelocked) {
+			return this.popupReply(`Unnamed users cannot challenge.`);
+		}
```
Comment (Zarel) [inline server/chat-commands/core.ts:1242]:
```suggestion
		if (!user.named) {
			return this.popupReply(`You must choose a username before you challenge someone.`);
		}
```

Namelocked users should be able to challenge other locked/namelocked users but no one else; this was discussed elsewhere.

**Comment by Zarel** [comment]
Nice! `Ladders.challenges.get(...)?.length` makes it a lot nicer!

File: server/chat-commands/core.ts (hunk 4 above)
Comment (mia-pi-git) [reply server/chat-commands/core.ts]: Should i add that here, as well?

Comment (Zarel) [reply server/chat-commands/core.ts]: New PR also works.

Comment (mia-pi-git) [reply server/chat-commands/core.ts]: Okay, then!
I had that committed so i guess i'll just push and PR.

**Pull request merged**